### PR TITLE
Support EPs that compile nodes in a minimal build.

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add support for EPs that compile nodes in a minimal build. This will enable the NNAPI EP to be used at runtime with an ORT format model.
  - add 'extended' minimal build that pulls in the infrastructure required to compile a node at runtime
    - there is a non-trivial amount of infrastructure required, so separating from the base minimal build where users may not want to pay the binary size price for features they do not need.
  - add ability to create compiled node without creating a new Model instance
    - creating a Model would result in pulling in all the Graph code, so we use a GraphViewer for the set of nodes to be compiled.
  - update graph partitioning to enable creating an ORT format model where nodes may be taken by a compiling EP, and to work in a minimal build with an ORT format model
  - provide mechanism to fallback to CPU EP for nodes that are not handled at runtime by a compiling EP
    - when creating the ORT format model the compiling EP provides information on the nodes it could potentially handle. we do not do further optimizations on those nodes so that they're guaranteed to be available to the compiling EP and included in the required kernels when creating the minimal build. at runtime the compiling EP may not be available, or the device may have limited support, so fewer nodes may be taken by the EP. in that case we use the CPU EP to execute any left over nodes. 
  - add test EP that can be used to unit test the behavior
    - compiling EPs tend to be device/hardware specific so having a CPU based EP that replicates the functionality makes testing significantly easier

**Motivation and Context**
Enable using NNAPI on device with an ORT format model.

### CODE REVIEW

File: onnxruntime/core/framework/session_state.cc [hunk 1]
```diff
@@ -113,26 +113,42 @@ void SessionState::CreateGraphInfo() {
...
-Status SessionState::PopulateKernelCreateInfo(KernelRegistryManager& kernel_registry_manager) {
+Status SessionState::PopulateKernelCreateInfo(KernelRegistryManager& kernel_registry_manager,
+                                              bool saving_ort_format) {
   for (auto& node : graph_.Nodes()) {
     const KernelCreateInfo* kci = nullptr;
-    ORT_RETURN_IF_ERROR(kernel_registry_manager.SearchKernelRegistry(node, &kci));
+
+    auto status = kernel_registry_manager.SearchKernelRegistry(node, &kci);
+    if (!status.IsOK() && saving_ort_format) {
+      // if we didn't find the kernel and are saving to ORT format an EP that compiles nodes is enabled.
```
Comment (gwang-msft) [inline onnxruntime/core/framework/session_state.cc:132]: Maybe and Node is not using CPU? If it's using CPU the search below will still fail, but why search twice?

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -1130,51 +1158,73 @@ common::Status InferenceSession::Initialize() {
...
-                                                  insert_cast_transformer_,
-                                                  *session_state_));
-
-    // now that all the transforms are done, call Resolve on the main graph. this will recurse into the subgraphs.
-    ORT_RETURN_IF_ERROR_SESSIONID_(graph.Resolve());
+    bool loading_ort_format = !ort_format_model_bytes_.empty();
+    bool saving_model = !session_options_.optimized_model_filepath.empty();
+    bool saving_ort_format = false;
+    if (saving_model) {
+      std::string model_type = session_options_.GetConfigOrDefault(kOrtSessionOptionsConfigSaveModelFormat, "");
```
Comment (gwang-msft) [inline onnxruntime/core/session/inference_session.cc:1165]: Do we want to support saving model when loading_ort_format?

File: onnxruntime/core/framework/session_state.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/framework/session_state.cc:132]: During partitioning, all nodes will generally be assigned to valid EPs. The only exception is when we assign them to a compiling EP but do not compile them. We should only ever satisfy the 'if' when that occurs, and the only option we have currently is to fallback to CPU for those nodes. 

So the second call to SearchKernelRegistry should never fail. 

---
In reply to: [522558208](https://github.com/microsoft/onnxruntime/pull/5776#discussion_r522558208) [](ancestors = 522558208)

File: onnxruntime/core/session/inference_session.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/session/inference_session.cc:1165]: Probably not. You could run the ORT model partitioning and compile nodes which results in something that can't be saved. 

I will address in a separate PR though. There was a github issue this week where someone was attempting to save a model that had compiled TensorRT nodes, so we can just do a more general check that there are no compiled kernels and disallow save in that case. 

With that check, it doesn't matter if someone attempts to load and save an ORT format model as it won't be changed given the only thing PartitionOrtFormatModel can do is compile kernels.

---
In reply to: [522558762](https://github.com/microsoft/onnxruntime/pull/5776#discussion_r522558762) [](ancestors = 522558762)

File: onnxruntime/core/graph/graph.cc [hunk 3]
```diff
@@ -133,21 +133,26 @@ static TypeProto TypeProtoFromTensorProto(const TensorProto& tensor) {
 
   return t;
 }
+#endif
```
Comment (gwang-msft) [inline onnxruntime/core/graph/graph.cc:136]: >endif [](start = 1, length = 5)

nit, #endif !defined(ORT_MINIMAL_BUILD), and some other places

File: onnxruntime/core/framework/graph_partitioner.cc [hunk 4]
```diff
@@ -41,6 +43,17 @@ NonCudaOps non_cuda;
...
 
+// minimal KernelDef based on MetaDef instead of a Function based node
+static KernelDefBuilder& BuildFusedKernelDef(KernelDefBuilder& builder, const IndexedSubGraph::MetaDef& metadef,
+                                             const std::string& provider_type) {
+  builder.SetName(metadef.name)
+      .SetDomain(metadef.domain)
+      .SinceVersion(metadef.since_version)
+      .Provider(provider_type);
+  return builder;
+}
```
Comment (gwang-msft) [inline onnxruntime/core/framework/graph_partitioner.cc:54]: This part is kind of confusing, why is it return the same ref back?

Comment (skottmckay) [reply onnxruntime/core/framework/graph_partitioner.cc:54]: The return value isn't used. Will change to return void.

Comment (skottmckay) [reply onnxruntime/core/framework/graph_partitioner.cc:54]: Note that we can't call Build() here to finalize as on one code path we call the kernel registry's Register method with the builder.

File: onnxruntime/test/providers/internal_testing/internal_testing_tests.cc [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+  // prepare outputs
+  std::vector<std::string> output_names;
+  output_names.push_back("Plus214_Output_0");
+  std::vector<OrtValue> fetches;
+
+  ASSERT_STATUS_OK(session.Run(feeds, output_names, &fetches));
+
+  if (custom_ep_enabled) {
+    // check that the output is all zeros. the dummy EP produces output of the correct shape will all zeros, so any
```
Comment (gwang-msft) [inline onnxruntime/test/providers/internal_testing/internal_testing_tests.cc:64]: >will [](start = 93, length = 4)

with?

**Comment by gwang-msft** [comment]
For all the internal_testing_execution_provider code, does it have to be in onnxruntime/core and expose in include?
Can we create the EP under onnxruntime/test and use it in test only?

**Comment by skottmckay** [comment]
> For all the internal_testing_execution_provider code, do these ave to be in onnxruntime/core and expose in include/...?
> Can we create the EP under onnxruntime/test and use it in test only?

Good idea. Should be able to. Will try and make it fully internal.

File: onnxruntime/test/util/test_utils.cc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+static int CountAssignedNodes(const Graph& current_graph, const std::string& ep_type) {
+  int count = 0;
+
+  for (const auto& node : current_graph.Nodes()) {
+    if (node.GetExecutionProviderType() == ep_type) {
+      ++count;
+    }
+
+    if (node.ContainsSubgraph()) {
```
Comment (gwang-msft) [inline onnxruntime/test/util/test_utils.cc:57]: >if [](start = 4, length = 2)

else if?
Seems these 2 cases are mutually exclusive

File: onnxruntime/test/util/test_utils.cc [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      }
+    }
+  }
+
+  return count;
+}
+
+void RunAndVerifyOutputsWithEP(const ORTCHAR_T* model_path, const char* log_id,
+                               std::unique_ptr<IExecutionProvider> execution_provider,
+                               const NameMLValMap& feeds) {
```
Comment (gwang-msft) [inline onnxruntime/test/util/test_utils.cc:69]: Do you want to add an bool to not compare output, but only CountAssignedNodes? For example running NNAPI on windows will pass CountAssignedNodes but fail the VerifyOutputs

File: onnxruntime/test/providers/internal_testing/internal_testing_tests.cc [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    EXPECT_EQ(supported_ops.count(node.OpType()), size_t(0))
+        << "Nodes with supported op types should have been replaced. Node with type " << node.OpType() << " was not.";
+    if (node.GetExecutionProviderType() == kInternalTestingExecutionProvider) {
+      NodeComputeInfo* compute_func = nullptr;
+      EXPECT_STATUS_OK(func_mgr.GetFuncs(node.Name(), compute_func));
+      EXPECT_NE(compute_func, nullptr);
+      ++count;
+    }
+
+    if (node.ContainsSubgraph()) {
```
Comment (gwang-msft) [inline onnxruntime/test/providers/internal_testing/internal_testing_tests.cc:191]: >if [](start = 4, length = 2)

else if?

File: onnxruntime/core/framework/graph_partitioner.cc [hunk 9]
```diff
@@ -231,12 +359,158 @@ static Status InlineNodes(Graph& graph, bool& modified_graph) {
...
+
+// Simplified partitioning where custom EPs may produce compiled nodes.
+// EPs with static kernels do not need to be processed as their kernels are matched via hash information serialized
+// as part of the ORT format model.
+Status GraphPartitioner::PartitionOrtFormatModel(
+    Graph& graph, FuncManager& func_mgr,
+    KernelRegistry& fused_kernel_registry,
+    std::unordered_map<std::string, uint64_t>& compiled_kernel_hashes) const {
+  int unique_id = 0;
+
```
Comment (gwang-msft) [inline onnxruntime/core/framework/graph_partitioner.cc:496]: Move this into GraphPartitioner::Partition? 
And rename to fused_node_id to make it clear?
Maybe some comments about this is unique per-model?

File: onnxruntime/test/util/test_utils.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/test/util/test_utils.cc:57]: Not necessarily. This could be used with any EP (e.g. CUDA) where the EP being tested can take a node with a subgraph.

File: onnxruntime/test/util/test_utils.cc (hunk 7 above)
Comment (skottmckay) [reply onnxruntime/test/util/test_utils.cc:69]: I considered that, but for the NNAPI test we're just going to load the model, so all the work to get the CPU EP outputs isn't needed either.

**Review by pranavsharma — COMMENTED** [review]
Can we add a short writeup for this in github for new developers (not necessarily working on mobile related stuff) as it's changing how partitioning is done? Probably the mobile workflow (partitioning, support for EPs, etc.) needs to be documented separately.

File: cmake/CMakeLists.txt [hunk 10]
```diff
@@ -120,6 +120,7 @@ option(onnxruntime_DISABLE_RTTI "Disable RTTI" OFF)
 # For now onnxruntime_DISABLE_EXCEPTIONS will only work with onnxruntime_MINIMAL_BUILD, more changes (ONNX, non-CPU EP, ...) are required to run this standalone
 option(onnxruntime_DISABLE_EXCEPTIONS "Disable exception handling. Requires onnxruntime_MINIMAL_BUILD currently." OFF)
 option(onnxruntime_MINIMAL_BUILD "Exclude as much as possible from the build. Support ORT format models. No support for ONNX format models." OFF)
+option(onnxruntime_EXTENDED_MINIMAL_BUILD "onnxruntime_MINIMAL_BUILD with support for execution providers that compile kernels." OFF)
```
Comment (pranavsharma) [inline cmake/CMakeLists.txt:123]: Would the scope of extended minimal build expand under the same option name? If not, suggest naming this specific to the functionality it is enabling.

File: include/onnxruntime/core/graph/constants.h [hunk 11]
```diff
@@ -37,49 +37,53 @@ constexpr const char* kMIGraphXExecutionProvider = "MIGraphXExecutionProvider";
 constexpr const char* kAclExecutionProvider = "ACLExecutionProvider";
 constexpr const char* kArmNNExecutionProvider = "ArmNNExecutionProvider";
 constexpr const char* kRocmExecutionProvider = "ROCMExecutionProvider";
-constexpr const char *providers_available[] = {
-  kCpuExecutionProvider,
+
+// EP used for internal testing
+constexpr const char* kInternalTestingExecutionProvider = "InternalTestingExecutionProvider";
```
Comment (pranavsharma) [inline include/onnxruntime/core/graph/constants.h:42]: Can we exclude a test EP name from this file? This file is for all the officially supported EPs.

File: include/onnxruntime/core/graph/graph.h [hunk 12]
```diff
@@ -578,6 +579,15 @@ class Graph {
   /** Gets the path of the owning model, if any. */
   const Path& ModelPath() const;
 
+  /** Returns true if this is a subgraph or false if it is a high-level graph. */
+  bool IsSubgraph() const { return parent_graph_ != nullptr; }
```
Comment (pranavsharma) [inline include/onnxruntime/core/graph/graph.h:583]: Many functions have been re-ordered in this file. Haven't looked very closely. I suspect it's to support the new ifdef?

File: onnxruntime/core/session/inference_session.h [hunk 13]
```diff
@@ -543,6 +542,11 @@ class InferenceSession {
   std::vector<std::string> transformers_to_enable_;
 #endif
 
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
+  Status PartitionOrtFormatModel(onnxruntime::Graph& graph, const ExecutionProviders& providers,
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.h:546]: Noticed that some functions use ORT_MUST_USE_RESULT. We're not consistent. Should new functions have this as well?

File: tools/ci_build/build.py [hunk 14]
```diff
@@ -50,6 +50,49 @@ def _check_python_version():
 _check_python_version()
 
 
+def _openvino_verify_device_type(device_read):
```
Comment (pranavsharma) [inline tools/ci_build/build.py:53]: unrelated change?

**Comment by skottmckay** [comment]
Yeah - I have some documentation changes in progress (will be separate PR) I can add that info to. 

---
In reply to: [531602927](https://github.com/microsoft/onnxruntime/pull/5776#issuecomment-531602927) [](ancestors = 531602927)

File: cmake/CMakeLists.txt (hunk 10 above)
Comment (skottmckay) [reply cmake/CMakeLists.txt:123]: Not totally clear. I was thinking it we were to add support for runtime optimization there would be some overlap and both could be considered 'extended' functionality. I'm also trying to keep the number of different #ifdef values in the code under control. For example if we added runtime optimization you're getting into something horrendous for Graph::AddNode of `#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_COMPILING_EP_MINIMAL_BUILD) || defined(ORT_RUNTIME_OPT_MINIMAL_BUILD)`. And add one more feature that overlaps and it gets worse.

So I went with something generic we can add to until we have a hard requirement to split these additional functionalities out separately.

File: include/onnxruntime/core/graph/graph.h (hunk 12 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/graph.h:583]: Re-ordering was to group things that fall under the same #ifdef's without moving them too far. Just trying to reduce the number of different ifdef blocks.

File: onnxruntime/core/session/inference_session.h (hunk 13 above)
Comment (skottmckay) [reply onnxruntime/core/session/inference_session.h:546]: Not sure what the policy should be. I thought we added that to the public APIs initially, and following that it seemed to bleed down a bit. I thought the VS Code Analysis could find issues where a return value is ignored, so it may not be necessary to add ORT_MUST_USE_RESULT to internal code.

File: tools/ci_build/build.py (hunk 14 above)
Comment (skottmckay) [reply tools/ci_build/build.py:53]: I can split this out if you want. I couldn't take the insanity of defining a random function in the middle of nowhere with a name that provided no context to what it did.

File: include/onnxruntime/core/graph/constants.h (hunk 11 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/constants.h:42]: I will move it to the internal framework/utils.h. It's used in ProviderIsCpuBased in framework/utils.cc which is used by partitioning logic, so it needs to exist at that level (or we need the EP API to have a property saying whether it's CPU based so we don't have hardcoded lists for that).

File: onnxruntime/test/providers/internal_testing/internal_testing_tests.cc (hunk 8 above)
Comment (skottmckay) [reply onnxruntime/test/providers/internal_testing/internal_testing_tests.cc:191]: In this case as currently implemented it can't be both, but they're conceptually separate tests so I don't want to couple them with an `else if`. The testing EP could change to test taking control flow nodes, at which point the `else if` would become invalid (which is a hint that we should keep them separate).

---
In reply to: [523839104](https://github.com/microsoft/onnxruntime/pull/5776#discussion_r523839104) [](ancestors = 523839104)

File: tools/ci_build/build.py (hunk 14 above)
Comment (skottmckay) [reply tools/ci_build/build.py:53]: https://github.com/microsoft/onnxruntime/pull/5822

File: cmake/CMakeLists.txt (hunk 10 above)
Comment (pranavsharma) [reply cmake/CMakeLists.txt:123]: Yeah, this is fine. More functionality will be covered under 'extended'. You answered my question.

File: include/onnxruntime/core/graph/constants.h (hunk 11 above)
Comment (pranavsharma) [reply include/onnxruntime/core/graph/constants.h:42]: Yeah, moving it to internal framework/utils.h is a better idea. Adding an interface to EPs raises more questions.

File: onnxruntime/test/providers/internal_testing/internal_testing_tests.cc [hunk 15]
```diff
@@ -0,0 +1,239 @@
+#include "core/common/logging/logging.h"
```
Comment (pranavsharma) [inline onnxruntime/test/providers/internal_testing/internal_testing_tests.cc:1]: License header

**Review by pranavsharma — APPROVED** [review]
